src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -205,57 +250,102 @@ bool UccDiscoveryPlugin::_dictionary_segments_contain_duplicates(const std::shar
 }
 
 template <typename ColumnDataType>
-bool UccDiscoveryPlugin::_uniqueness_holds_across_segments(const std::shared_ptr<const Table>& table,
-                                                           const ColumnID column_id) {
+bool UccDiscoveryPlugin::_uniqueness_holds_across_segments(
+    const std::shared_ptr<const Table>& table, const std::string& table_name, const ColumnID column_id,
+    const std::shared_ptr<TransactionContext>& transaction_context,
+    const std::optional<TableKeyConstraint>& existing_ucc) {
+  // `distinct_values_across_segments` collects the segment values from all chunks.
+  auto distinct_values_across_segments = std::unordered_set<ColumnDataType>{};
+  auto unmodified_chunks = std::vector<ChunkID>();
+
   const auto chunk_count = table->chunk_count();
-  // `distinct_values` collects the segment values from all chunks.
-  auto distinct_values = std::unordered_set<ColumnDataType>{};
 
+  // Iterate through all unmodified chunks and collect their distinct values. If a segment does not add the expected
+  // number of distinct values, we know that a duplicate exists in the segment / column and can return early.
   for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
     const auto source_chunk = table->get_chunk(chunk_id);
     if (!source_chunk) {
+      // If this chunk does not exist, we do not need to check it with the validate operator later either.
+      unmodified_chunks.push_back(chunk_id);
       continue;
     }
     const auto source_segment = source_chunk->get_segment(column_id);
-    if (!source_segment) {
-      continue;
+    // Was the chunk modified since the last UCC validation (inserted or updated rows)? If not, than we do not need to
+    // probe its values for duplicates later, because duplicates can only be introduced by chunks that were modified
+    // after the last validation.
+    const bool not_modified_since_last_validation =
+        existing_ucc && existing_ucc->last_validation_result() == ValidationResultType::VALID &&
+        source_chunk->mvcc_data()->max_begin_cid.load() <= existing_ucc->last_validated_on();
+    if ((source_chunk->invalid_row_count() == 0 && !source_chunk->is_mutable()) || not_modified_since_last_validation) {
+      // The set of distinct values across all segments should grow by the number of rows in the segment because,
+      // if it does not, it means some value must have been inserted twice -> duplicate detected.
+      // In this case, the UCC is violated.
```

Comment:
```suggestion
      // The set of distinct values across all segments should grow by the number of rows in the segment. Otherwise,
      // some value must have been inserted twice, and we detected a duplicate that violates the UCC.
```